optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -851,20 +857,41 @@ public void validateConstraintWeight(String constraintPackage, String constraint
     // Extraction methods
     // ************************************************************************
 
+    public void visitAllFacts(Solution_ solution, Consumer<Object> visitor) {
```

Comment:
Replaces `getAllFacts()`.

Consider the following use case, often found in score directors:

    List facts = getAllFacts();
    facts.forEach(fact -> session.insert(fact));

This iterates twice over the list - first to create it, then to use it to insert facts into a session. At the end, it throws away the list as a garbage. All of this is inefficient.

This new mechanism only ever iterates once, and never creates any garbage.